server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java

```diff
@@ -121,6 +121,9 @@ private ByteBuffer toByteBuffer(byte[] bytes) {
 
     private byte[] toByteArray(Row row) {
         ByteBuffer byteBuffer = row.getBytes(DATA);
+        if (byteBuffer.hasArray()) {
+            return byteBuffer.array();
```

Comment:
there's no guarantee that:
* the buffer won't be recycled by another layer
* the array has the size of the data

Reply:
We likely can drop this commit if you wish.

Reply:
Dropped

Reply:
BTW you are wrong:

https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#hasArray()

```
Tells whether or not this buffer is backed by an accessible byte array.

If this method returns true then the array and arrayOffset methods may safely be invoked. 
```

https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#array()

```
Returns the byte array that backs this buffer  (optional operation). 
```

As I understand that you feel concerned I will open a separated PR, But still want this code merged.

Reply:
see https://github.com/apache/james-project/pull/316

Reply:
Let's say I'm wrong (because you look really sure of your opinion). Then open ByteBuffer.wrap and tell me what happen when offset/limit is used. Do you see any shrinking of the buffer somewhere?

I can remember being bitten by this in the past, please look closer to this problem because I'm confident it's not correct code.

Reply:
To be fair, from what I see we could use the offset and the length to generate a stream on demand without an array copy.

Here the callers need a byte[] access.